## Working through Robert's Notebook:

### Imports:

In [3]:
import pandas as pd

In [4]:
parent_dir = '../'

### EDA:

In [5]:
df_raw = pd.read_csv(parent_dir + 'data/weather_madrid_LEMD_1997_2015.csv')
df_raw.head()

CET  Max TemperatureC  Mean TemperatureC  Min TemperatureC  \
0  1997-1-1               7.0                4.0               2.0   
1  1997-1-2               7.0                3.0               0.0   
2  1997-1-3               5.0                3.0               2.0   
3  1997-1-4               7.0                3.0              -1.0   
4  1997-1-5               2.0                0.0              -1.0   

   Dew PointC  MeanDew PointC  Min DewpointC  Max Humidity   Mean Humidity  \
0         5.0             3.0            2.0         100.0            95.0   
1         6.0             3.0            0.0         100.0            92.0   
2         5.0             1.0           -1.0         100.0            85.0   
3        -2.0            -3.0           -4.0          86.0            63.0   
4         2.0             0.0           -3.0         100.0            95.0   

    Min Humidity  ...   Max VisibilityKm   Mean VisibilityKm  \
0           76.0  ...               10.0                 9.0   
1           71.0  ...               10.0                 9.0   
2           70.0  ...               10.0                10.0   
3           49.0  ...               10.0                10.0   
4           86.0  ...               10.0                 5.0   

    Min VisibilitykM   Max Wind SpeedKm/h   Mean Wind SpeedKm/h  \
0                4.0                   13                     6   
1                4.0                   26                     8   
2                7.0                   27                    19   
3               10.0                   27                    19   
4                1.0                   14                     6   

    Max Gust SpeedKm/h  Precipitationmm   CloudCover     Events  \
0                  NaN              0.0          6.0        NaN   
1                 47.0              0.0          5.0       Rain   
2                  NaN              0.0          6.0  Rain-Snow   
3                 40.0              0.0          2.0        NaN   
4                  NaN              0.0          7.0       Snow   

   WindDirDegrees  
0             229  
1             143  
2             256  
3             284  
4               2  

[5 rows x 23 columns]

Update the column names

In [6]:
rename_dict =  {'CET'                           :'date',
                'Max TemperatureC'              :'temperature_max',
                'Mean TemperatureC'             :'temperature',
                'Min TemperatureC'              :'temperature_min',
                'Dew PointC'                    :'dewpoint_max',
                'MeanDew PointC'                :'dewpoint',
                'Min DewpointC'                 :'dewpoint_min',
                'Max Humidity'                  :'humidity_max',
                ' Mean Humidity'                :'humidity',
                ' Min Humidity'                 :'humidity_min',
                ' Max Sea Level PressurehPa'    :'pressure_max',
                ' Mean Sea Level PressurehPa'   :'pressure',
                ' Min Sea Level PressurehPa'    :'pressure_min',
                ' Max VisibilityKm'             :'visibility_max',
                ' Mean VisibilityKm'            :'visibility',
                ' Min VisibilitykM'             :'visibility_min',
                ' Max Wind SpeedKm/h'           :'windspeed_max',
                ' Mean Wind SpeedKm/h'          :'windspeed',
                ' Max Gust SpeedKm/h'           :'gustspeed',      
                'Precipitationmm'               :'precipitation',
                ' CloudCover'                   :'cloudcover',
                ' Events'                       :'events',
                'WindDirDegrees'                :'winddirection',
                }

In [8]:
df_raw.rename(columns = rename_dict, inplace = True)

In [9]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6812 entries, 0 to 6811
Data columns (total 23 columns):
date               6812 non-null object
temperature_max    6810 non-null float64
temperature        6809 non-null float64
temperature_min    6810 non-null float64
dewpoint_max       6810 non-null float64
dewpoint           6810 non-null float64
dewpoint_min       6810 non-null float64
humidity_max       6810 non-null float64
humidity           6810 non-null float64
humidity_min       6810 non-null float64
pressure_max       6812 non-null int64
pressure           6812 non-null int64
pressure_min       6812 non-null int64
visibility_max     5872 non-null float64
visibility         5872 non-null float64
visibility_min     5872 non-null float64
windspeed_max      6812 non-null int64
windspeed          6812 non-null int64
gustspeed          3506 non-null float64
precipitation      6812 non-null float64
cloudcover         5440 non-null float64
events             1798 non-null object
win

Convert date column to datetime object

In [11]:
df_raw['date'] = pd.to_datetime(df_raw['date'])

In [14]:
type(df_raw.date[0])

pandas._libs.tslibs.timestamps.Timestamp

Check the current shape of the df:

In [16]:
df_raw.shape

(6812, 23)

The group decided that we would use the data after 2013 as the test data so we seperate that out here in order to not affect model design or exploration:

In [18]:
(df,test_df) = (df_raw[df_raw.date.dt.year < 2013], df_raw[df_raw.date.dt.year >= 2013])

In [19]:
df.shape

(5717, 23)

In [20]:
test_df.shape

(1095, 23)

In [22]:
df.describe().T

count         mean         std    min     25%     50%  \
temperature_max  5715.0    20.875941    8.841317    0.0    13.0    20.0   
temperature      5714.0    14.519776    7.565159   -3.0     8.0    14.0   
temperature_min  5715.0     8.587402    6.832758  -10.0     3.0     9.0   
dewpoint_max     5715.0     8.030796    4.783309  -12.0     5.0     8.0   
dewpoint         5715.0     4.902012    4.694278  -15.0     2.0     5.0   
dewpoint_min     5715.0     1.415923    4.958462  -22.0    -2.0     2.0   
humidity_max     5715.0    81.137533   17.396003   16.0    68.0    87.0   
humidity         5715.0    58.456868   19.629104   15.0    42.0    59.0   
humidity_min     5715.0    36.221347   19.221618    4.0    20.0    33.0   
pressure_max     5717.0  1020.398461    6.068196  996.0  1017.0  1020.0   
pressure         5717.0  1018.003848    6.394319  990.0  1014.0  1018.0   
pressure_min     5717.0  1015.529823    6.801470  965.0  1012.0  1016.0   
visibility_max   4777.0    11.010467    4.675784    1.0    10.0    10.0   
visibility       4777.0     9.919824    3.066143    0.0    10.0    10.0   
visibility_min   4777.0     8.504919    3.895358    0.0     7.0    10.0   
windspeed_max    5717.0    21.658037    9.784401    0.0    13.0    21.0   
windspeed        5717.0     9.006297    5.081384    0.0     5.0     8.0   
gustspeed        2682.0    43.663311   12.337341   19.0    35.0    42.0   
precipitation    5717.0     0.020388    0.507565    0.0     0.0     0.0   
cloudcover       4630.0     3.239525    1.792508    0.0     2.0     3.0   
winddirection    5717.0   195.519853  120.985340   -1.0    60.0   223.0   

                    75%     max  
temperature_max    29.0    41.0  
temperature        21.0    32.0  
temperature_min    14.0    28.0  
dewpoint_max       12.0    20.0  
dewpoint            8.0    16.0  
dewpoint_min        5.0    14.0  
humidity_max       94.0   100.0  
humidity           75.0   100.0  
humidity_min       49.0   100.0  
pressure_max     1024.0  1041.0  
pressure         1022.0  1040.0  
pressure_min     1020.0  1038.0  
visibility_max     10.0    31.0  
visibility         10.0    31.0  
visibility_min     10.0    31.0  
windspeed_max      27.0   182.0  
windspeed          11.0    39.0  
gustspeed          52.0   103.0  
precipitation       0.0    32.0  
cloudcover          5.0     8.0  
winddirection     299.0   360.0

In [24]:
df.date.min()

Timestamp('1997-01-01 00:00:00')

In [25]:
df.date.max()

Timestamp('2012-12-31 00:00:00')

In [26]:
dr = pd.date_range(df.date.min(), df.date.max())

In [29]:
date_range = pd.Series(dr)

In [30]:
date_range

0      1997-01-01
1      1997-01-02
2      1997-01-03
3      1997-01-04
4      1997-01-05
          ...    
5839   2012-12-27
5840   2012-12-28
5841   2012-12-29
5842   2012-12-30
5843   2012-12-31
Length: 5844, dtype: datetime64[ns]

Now that we have a date range, we can check to see if there are any days missing from the dataset:

In [31]:
len(df)

5717

In [32]:
len(date_range)

5844

In [33]:
len(date_range) - len(df)

127

So we can see there are 127 days missing from the dataframe.  

Let's also check to see if there are any days that don't have temperatures:

In [39]:
len(df) - df.temperature.count()

3

Or you could also do:

In [41]:
df.temperature.isnull().sum()

3

So there are 3 days that don't have temperature data.

Let's check the test data to see if there's anything missing:

In [43]:
test_dr = pd.Series(pd.date_range(test_df.date.min(), test_df.date.max()))